### ISSUE TITLE
Segfault during device_map="auto" weight materialization on RTX 5090 D v2 (GB202-240, sm_120)

### ISSUE THREAD

**Issue by XHXIAIEIN** [pytorch/pytorch#175614]
## Bug Description

Loading a model via `AutoModelForCausalLM.from_pretrained()` with `device_map="auto"` (which delegates to `accelerate`) causes a **Segmentation Fault (exit code 139)** on the **NVIDIA GeForce RTX 5090 D v2**, a China-market Blackwell variant using the **GB202-240** die with compute capability **sm_120**.

Basic CUDA operations work correctly — `torch.randn`, `torch.matmul` on bf16 tensors, and even direct `safetensors.torch.load_file(..., device='cuda:0')` all succeed. The crash occurs specifically during the weight dispatch/materialization phase orchestrated by `accelerate` when `device_map="auto"` is used.

The crash happens early in the process (around layers 0–8 of the model), ruling out an OOM condition. It reproduces with both bf16 loading and 4-bit quantization via `bitsandbytes`.

### Why this matters

The RTX 5090 D v2 (GB202-240) is a **China-specific variant** that differs from the standard RTX 5090 (GB202-300) in die configuration and VRAM. It is unlikely to be present in the PyTorch CI/CD test matrix, but it shares `sm_120` compute capability with other Blackwell GPUs. This may indicate a broader issue with Blackwell support in the weight dispatch codepath, or it may be specific to this particular SKU's hardware characteristics.

### Hardware Comparison

| Attribute | RTX 5090 (Standard) | RTX 5090 D v2 (China) |
|---|---|---|
| **Die** | GB202-300 | GB202-240 |
| **VRAM** | 32 GB GDDR7 | 24 GB GDDR7 |
| **Memory Bus** | 512-bit | 384-bit |
| **Compute Capability** | sm_120 (12.0) | sm_120 (12.0) |
| **Market** | Global | China-specific |

## Minimal Reproducible Example

```python
import torch
from transformers import AutoModelForCausalLM

print(f"PyTorch: {torch.__version__}")
print(f"CUDA available: {torch.cuda.is_available()}")
print(f"GPU: {torch.cuda.get_device_name(0)}")
print(f"Compute capability: {torch.cuda.get_device_capability(0)}")

# ============================================================
# TEST 1: Basic CUDA ops — PASSES
# ============================================================
a = torch.randn(4096, 4096, dtype=torch.bfloat16, device="cuda:0")
b = torch.randn(4096, 4096, dtype=torch.bfloat16, device="cuda:0")
c = torch.matmul(a, b)
print(f"matmul result shape: {c.shape}, device: {c.device}")  # OK

# ============================================================
# TEST 2: Direct safetensors load to GPU — PASSES
# ============================================================
from safetensors.torch import load_file
tensors = load_file("model-00001-of-00008.safetensors", device="cuda:0")
print(f"Loaded {len(tensors)} tensors directly to GPU")  # OK: 41 tensors, ~3.58 GB

# ============================================================
# TEST 3: device_map="auto" — SEGFAULT
# ============================================================
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-14B",
    device_map="auto",
    torch_dtype=torch.bfloat16,
)
# ^^^ Process crashes with Segmentation Fault (exit code 139)
#     Crash occurs during weight materialization, around layers 0-8
```

### Additional reproduction attempts (all crash)

```python
# 4-bit quantization — same Segfault
from transformers import BitsAndBytesConfig
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-14B",
    device_map="auto",
    quantization_config=BitsAndBytesConfig(load_in_4bit=True),
)

# Explicit max_memory — same Segfault
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-14B",
    device_map="auto",
    torch_dtype=torch.bfloat16,
    max_memory={0: "20GiB", "cpu": "30GiB"},
)
```

## Expected Behavior

`AutoModelForCausalLM.from_pretrained(..., device_map="auto")` should load the model onto the GPU (and CPU/disk overflow if needed) without crashing, consistent with the behavior on other CUDA-capable GPUs.

## Actual Behavior

The process terminates with a **Segmentation Fault (exit code 139)** during the weight dispatch phase. No Python traceback is produced — the process is killed by the OS signal. The crash occurs early in model loading (layers 0–8), not at an OOM boundary.

## Environment

| Component | Version |
|---|---|
| **PyTorch** | `2.12.0.dev20260223+cu128` (nightly) |
| **CUDA (PyTorch)** | 12.8 |
| **NVIDIA Driver** | 591.86 |
| **Python** | 3.14 |
| **transformers** | latest (pip) |
| **accelerate** | 1.12.0 |
| **bitsandbytes** | 0.49.2 |
| **safetensors** | latest (pip) |
| **OS** | Windows 11 Pro (10.0.26200) |
| **GPU** | NVIDIA GeForce RTX 5090 D v2 (GB202-240, 24 GB, 384-bit) |
| **Compute Capability** | 12.0 (sm_120) |

## Suspected Root Cause

1. **`accelerate` dispatch codepath**: Since direct `safetensors` loading to CUDA works, the issue likely lies in how `accelerate` materializes and dispatches weights (possibly in `accelerate.utils.modeling.set_module_tensor_to_device` or related functions).

2. **sm_120 kernel coverage**: Although `cu128` should support sm_120, there may be a missing or broken kernel variant that is only triggered during the `accelerate` weight dispatch path.

3. **GB202-240 vs GB202-300**: The reduced memory bus width (384-bit vs 512-bit) and VRAM (24 GB vs 32 GB) may interact with memory allocation strategies in unexpected ways.

## Related

- This may also be relevant to `accelerate` (`huggingface/accelerate`); cross-filing if triage determines the fault is in the dispatch layer rather than a CUDA/driver issue.

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia

**Comment by XHXIAIEIN** [comment]
## Update: Confirmed Workaround — Manual CPU Load + `.to("cuda")`

After further testing, I found a reliable workaround that bypasses the segfault while still achieving full GPU inference:

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

# ✅ WORKS: Load to CPU first, then move to GPU manually
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-8B",
    dtype=torch.bfloat16,
    low_cpu_mem_usage=True,  # Important: avoids doubling RAM usage
)
model = model.to("cuda")  # Manual GPU transfer — no segfault

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-8B")
```

### End-to-End Verification

Successfully ran a full RAG pipeline (vector retrieval + LLM generation + self-reflection verification) with this approach:

| Metric | Value |
|---|---|
| Model | Qwen/Qwen3-8B (bf16, ~16 GB) |
| VRAM Usage | 15.3 GB / 24 GB |
| Inference Speed | ~8.9 tok/s |
| Generation Quality | Correct responses with high confidence |
| Self-Reflection | Passed reliability verification |

### What This Tells Us

- **GPU hardware is fully functional** — CUDA matmul, safetensors direct load, and manual `.to("cuda")` all work perfectly on RTX 5090 D v2 (sm_120)
- **The bug is specifically in the `device_map="auto"` dispatch path** — likely in how `accelerate` materializes meta-tensors or copies weight data during the sharded loading process
- **`low_cpu_mem_usage=True` on CPU followed by `.to("cuda")` uses a different code path** that avoids the problematic dispatch logic entirely

### Limitation

This workaround only works for models that fit entirely in VRAM (24 GB for this card). Models requiring CPU-GPU split (e.g., Qwen3-14B at ~28 GB bf16) still need `device_map="auto"`, which segfaults. Quantization (`load_in_4bit`) also segfaults because it also relies on `device_map`.

### Environment (same as original report)
- PyTorch: `2.12.0.dev20260223+cu128` (nightly)
- GPU: NVIDIA GeForce RTX 5090 D v2 (GB202-240, sm_120, 24 GB)
- Driver: 591.86
- OS: Windows 11 Pro

**Comment by XHXIAIEIN** [comment]
## Correction & Additional Verification

### 1. Version Corrections

The original report listed some versions imprecisely:

| Field | Originally reported | Actual |
|---|---|---|
| Python | 3.14 | **3.14.3** |
| transformers | "latest (pip)" | **5.2.0** |
| safetensors | "latest (pip)" | **0.7.0** |

### 2. Definitive OOM Exclusion Test

To rule out the possibility that the segfault was caused by insufficient memory (the original test used Qwen3-14B at ~28 GB bf16 which exceeds the 24 GB VRAM), I re-tested with **Qwen3-8B (~16 GB bf16)** which fits entirely in 24 GB VRAM:

```
VRAM: 20.2 / 23.9 GB free
RAM:  30.6 GB available

device_map="auto" with Qwen3-8B → Segmentation fault (exit code 139)
Crashed at: Loading weights: 19% | 76/399 | model.layers.6.self_attn.o_proj.weight
```

**This definitively rules out OOM.** The model (16 GB) had 20 GB VRAM free and 30 GB RAM free, yet still segfaulted at only 19% of weight loading. The crash is in the `accelerate` weight materialization path, not a memory boundary issue.

### 3. Clarification on 4-bit Quantization Case

The original report grouped the 4-bit (`bitsandbytes`) segfault with the bf16 `device_map="auto"` segfault. These **may be separate issues**:

- **bf16 + `device_map="auto"`**: Confirmed `accelerate` dispatch bug (proven above with Qwen3-8B)
- **4-bit + `device_map="auto"`**: Could be the same accelerate bug, OR a `bitsandbytes` kernel issue on sm_120. bitsandbytes 0.49.2 with cu128 does include sm_120 targets, but the kernels may not be fully validated on Blackwell. These should be treated as potentially separate failure modes.
